serve/mlc_serve/engine/engine_common.py

```diff
@@ -392,16 +438,28 @@ def evict_request(self, cancell_callback: Callable[[RequestId], None]) -> int:
                 candidate_victims = parallel_sample_requests
 
             request_to_remove = min(candidate_victims, key=lambda s: s.num_total_tokens)
-
-            # TODO(masahi): Properly support evicting a multi-sequence request
-            if self.current_batch[request_to_remove.request_id].num_sequences != 1:
-                cancell_callback(request_to_remove.request_id)
-                self.remove_request_from_batch(request_to_remove.request_id)
-                LOG.warn(
-                    "Preempting a multi-sequence request is currently not supported,"
-                    f" cancelling request '{request_to_remove.request_id}'",
+            victim_state = self.current_batch[request_to_remove.request_id]
+
+            if victim_state.num_sequences != 1:
+                prev_generated_token_counts = sum(
+                    [
+                        len(gen_seq.generated_token_ids)
+                        for gen_seq in victim_state.generation_sequences
+                    ]
                 )
-                continue
+                # We could allow evicting and restoring a parallel-sampling request whose prev_generated_token_counts
+                # is > max_num_batched_tokens, by making the model split a list of EvalMultiQuery requests into parts,
+                # so that an inference on each part can be done with the max_num_batched_tokens budget.
+                # But this introduces an undesirable coupling between the engine and the model.
+                if prev_generated_token_counts >= self.max_num_batched_tokens:
+                    cancell_callback(request_to_remove.request_id)
+                    self.remove_request_from_batch(request_to_remove.request_id)
+                    LOG.warn(
+                        f"Cancelling a parallel-sampling request '{request_to_remove.request_id}'"
+                        f"since it has generated more than {self.max_num_batched_tokens} tokens in total"
+                         "and currently we do not support preempting such request.",
+                    )
+                    continue
```

Comment:
@sunggg @elvin-n Please be aware of this limitation. Due to this, there is still a case when a parallel-sampling request is cancelled rather than preempted. 

In general, we don't have a good solution for preemption for a request which has generated more than `max_num_batched_tokens`. See also https://github.com/octoml/mlc-llm/issues/163. The easiest solution would be to stop generation at `max_num_batched_tokens`, but then we cannot support "unlimited" generation.